synapse/storage/devices.py

```diff
@@ -84,35 +85,99 @@ def get_devices_by_remote(self, destination, from_stream_id):
+            len(updates) > limit and
+            updates[-1][2] == updates[-2][2]
+        ):
+            # If so, cap our maximum stream_id at that final stream_id
+            stream_id_cutoff = updates[-1][2]
+
+        # Perform the equivalent of a GROUP BY
+        #
+        # Iterate through the updates list and copy non-duplicate
+        # (user_id, device_id) entries into a map, with the value being
+        # the max stream_id across each set of duplicate entries
+        #
+        # maps (user_id, device_id) -> stream_id
+        # as long as their stream_id does not match that of the last row
+        query_map = {}
+        for update in updates:
+            if update[2] >= stream_id_cutoff:
+                # Stop processing updates
+                break
+
+            key = (update[0], update[1])
+            query_map[key] = max(query_map.get(key, 0), update[2])
+
+        # If we ended up not being left over with any device updates to send
+        # out (because there was more device updates with the same stream_id
+        # that our defined limit allows), then just skip this stream_id.
+        #
+        # The list of updates associated with this stream_id is too large and
+        # thus we're just going to assume it was a client-side error and not
+        # send them. We return an empty list of updates instead.
+        if not query_map:
+            defer.returnValue((now_stream_id + 1, []))
+
+        results = yield self.runInteraction(
+            "_get_device_update_edus_by_remote_txn",
+            self._get_device_update_edus_by_remote_txn,
+            destination,
+            from_stream_id,
+            now_stream_id,
+            query_map,
+            limit,
+        )
+
+        defer.returnValue((now_stream_id, results))
+
     def _get_devices_by_remote_txn(
-        self, txn, destination, from_stream_id, now_stream_id
+        self, txn, destination, from_stream_id, now_stream_id, limit
     ):
+        """Return device update information for a given remote destination"""
         sql = """
-            SELECT user_id, device_id, max(stream_id) FROM device_lists_outbound_pokes
+            SELECT user_id, device_id, stream_id FROM device_lists_outbound_pokes
             WHERE destination = ? AND ? < stream_id AND stream_id <= ? AND sent = ?
-            GROUP BY user_id, device_id
-            LIMIT 20
+            ORDER BY stream_id
+            LIMIT ?
         """
-        txn.execute(sql, (destination, from_stream_id, now_stream_id, False))
+        txn.execute(sql, (destination, from_stream_id, now_stream_id, False, limit))
 
-        # maps (user_id, device_id) -> stream_id
-        query_map = {(r[0], r[1]): r[2] for r in txn}
-        if not query_map:
-            return (now_stream_id, [])
-
-        if len(query_map) >= 20:
-            now_stream_id = max(stream_id for stream_id in itervalues(query_map))
+        return list(txn)
 
+    def _get_device_update_edus_by_remote_txn(
```

Comment:
does this need to be a `txn` function? if you make it do `runInteraction` itself, we can free up the db connection for the second half of the function.